### PR TITLE
[training] feat: answer_only_vocab — auto-slice output projection for…

### PR DESCRIPTION
**Pull request by cuichenx**
… faster SFT

When fine-tuning with answer_only_loss, only a small fraction of vocabulary tokens appear in answers. This adds answer_only_vocab to FinetuningDatasetConfig which scans the training dataset, collects the N unique answer token IDs, and patches the output layer to compute logits over [N, H] instead of [V, H], reducing output-matmul and cross-entropy cost proportionally.

Key changes:
- src/megatron/bridge/training/vocab_slice.py: new module with collect_active_vocab_ids, install_vocab_slice, create_vocab_sliced_forward_step. Includes PP guard (skip non-post_process stages) and TP guard (warn+skip when TP>1 since weight is a local shard).
- src/megatron/bridge/training/config.py: add answer_only_vocab (bool) and answer_only_vocab_max_samples (Optional[int]) to FinetuningDatasetConfig.
- src/megatron/bridge/training/finetune.py: wire answer_only_vocab via _on_train_ds hook so slicing is installed after dataset build and before training, without touching pretrain.py's public API.
- src/megatron/bridge/training/pretrain.py: add generic _on_train_ds hook and fix closure rebinding bug in _capturing_provider (was infinite recursion).
- docs/training/answer-only-vocab.md: new doc covering when to use, loss change warning, parallelism support matrix, config example, and low-level API.
- tests/unit_tests/training/test_vocab_slice.py: 24 unit tests including PP/TP guard behavior.
- tests/functional_tests/test_groups/training/test_answer_only_vocab_parallel.py: multi-GPU smoke tests for PP=2, TP=2, CP=2 (all passing on 2x H100).

Measured speedup (60M model, V=64k -> N=260): 5.41s -> 3.03s/step (1.79x).

Warning: this changes the loss value (CE normalizes over N not V). See docs.

# What does this PR do ?

Add a one line overview of what this PR aims to accomplish.

# Changelog

- Add specific line by line info of high level changes in this PR.

# GitHub Actions CI

See the [CI section](https://github.com/NVIDIA-NeMo/Megatron-Bridge/blob/main/CONTRIBUTING.md#-running-github-ci)in the Contributing doc for how to trigger the CI. A Nvidia developer will need to approve and trigger the CI for external contributors.

# Before your PR is "Ready for review"

**Pre checks**:

- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA-NeMo/Megatron-Bridge/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?

If you haven't finished some of the above items you can still open "Draft" PR.

# Additional Information

- Picks up #2522

### CODE REVIEW

**Comment by cuichenx** [comment]
/ok to test https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/3679/commits/94720d10a404da2f95f1ab94e058b39549ffb4ca

File: src/megatron/bridge/training/vocab_slice.py [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+    # TP: the output layer weight is a local shard [V//TP, H].  Indexing it
+    # with global active_ids would silently produce wrong logits on ranks > 0.
+    try:
+        from megatron.core import parallel_state
+
+        tp_size = parallel_state.get_tensor_model_parallel_world_size()
+    except Exception:
+        tp_size = 1
+    if tp_size > 1:
+        logger.warning(
```
Comment (yaoyu-33) [inline src/megatron/bridge/training/vocab_slice.py:178]: raise error here. do not fallback

Comment (yaoyu-33) [reply src/megatron/bridge/training/vocab_slice.py:178]: can we just make padding so the tp works?

File: src/megatron/bridge/training/pretrain.py [hunk 2]
```diff
@@ -125,6 +135,22 @@ def _pretrain(
 
     config = state.cfg
     dataset_provider = get_dataset_provider(config.dataset)
+
+    # If a caller (e.g. finetune) needs the raw train_ds after it is built but
```
Comment (yaoyu-33) [inline src/megatron/bridge/training/pretrain.py:139]: make this a callback? and pass in via callback manager?

File: src/megatron/bridge/training/finetune.py [hunk 3]
```diff
@@ -50,4 +51,19 @@ def finetune(
     assert config.checkpoint.pretrained_checkpoint is not None or config.checkpoint.load is not None, (
         "Finetuning requires a loading from a pretrained checkpoint or resuming from a checkpoint"
     )
-    return pretrain(config, forward_step_func, callbacks=callbacks)
+
+    on_train_ds = None
```
Comment (yaoyu-33) [inline src/megatron/bridge/training/finetune.py:55]: similar, make this a callback

**Comment by cuichenx** [comment]
Architecture-only benchmarks on real production models (H100, BF16, recipe-default parallelism, 1% of vocab active, separate
  fresh processes per condition with proper warmup):

  | Model | Layout | Speedup |
  |---|---|---|
  | Qwen3-8B (dense) | TP=4, PP=1, seq=2048 | ~1.04× |
  | Qwen3-30B-A3B (MoE) | TP=4, PP=1, EP=4, seq=2048 | ~1.01× |
  | Qwen3-30B-A3B (MoE) | TP=4, PP=2, EP=4, mbs=8 | ~1.00× |
  | Qwen3-30B-A3B (MoE) | TP=4, PP=2, EP=4, mbs=32 | ~0.99× |

  All within run-to-run noise. **No real production model occupies the regime where this feature pays off.**

  The historical number was on a 60M toy (`H=512`, `V=64k`, tied embeddings) where the output projection alone was ~55% of total
  parameters. In Qwen3-8B that fraction is ~8%, and at TP=4 it drops to ~2% of per-rank step time. Slicing 99% of a 2% fraction is
  invisible.
